src/Generator/Passes/DelegatesPass.cs

```diff
@@ -119,61 +240,112 @@ private void AddDelegatesToDictionary(Declaration decl, Module module)
                                 }))),
                     Namespace = namespaceDelegates,
                     IsSynthetized = true,
-                    Access = decl is Method ? AccessSpecifier.Private : AccessSpecifier.Public
+                    Access = access
                 };
 
+            var typeDefType = new TypedefType
+            {
+                Declaration = @delegate
+            };
+
             var delegateString = @delegate.Visit(TypePrinter).Type;
             var existingDelegate = GetExistingDelegate(
                 module, delegateString);
 
             if (existingDelegate != null)
             {
-                Context.Delegates.Add(decl, existingDelegate);
-                return;
+                // Code to ensure that if a delegate is used for a virtual as well as something else, it finally ends up as public
+                if (existingDelegate.Access == AccessSpecifier.Private && access == AccessSpecifier.Public)
+                {
+                    existingDelegate.Access = AccessSpecifier.Public;
+                    if (libsDelegates.ContainsKey(module))
+                    {
+                        if (libsDelegates[module].ContainsKey(delegateString))
+                            libsDelegates[module].Remove(delegateString);
+                        libsDelegates[module].Add(delegateString, existingDelegate);
+                    }
+                    foreach (var dec in namespaceDelegates.Declarations.Where(m => m.Name == @delegate.Name))
+                    {
+                        ((TypedefDecl)dec).Access = AccessSpecifier.Public;
+                        break;
+                    }
+                }
+                // Code to check if there is an existing delegate with a different calling convention
+                // Add the new delegate with calling convention appended to it's name
+                if (existingDelegate.CallingConvention != callingConvention)
+                {
+                    delegateString += callingConvention.ToString();
+                    existingDelegate = new DelegateDefinition(module.OutputNamespace,
+                        delegateString, access, callingConvention);
+                    @delegate.Name += callingConvention.ToString();
```

Comment:
I assume this changes the name of the existing delegate. It's not necessary, they just need to be different. If one is called MyDelegate, and the other MyDelegateStd, it's enough, no need for the first one to be renamed to MyDelegateC.